app/js/DockerRunner.js

```diff
@@ -407,19 +408,28 @@ module.exports = DockerRunner = {
           })
         }
       )
-    return container.inspect(function(error, stats) {
-      if ((error != null ? error.statusCode : undefined) === 404) {
-        return createAndStartContainer()
-      } else if (error != null) {
-        logger.err(
-          { container_name: name, error },
-          'unable to inspect container to start'
-        )
-        return callback(error)
-      } else {
-        return startExistingContainer()
-      }
-    })
+    var inspectContainer = (isRetry) =>
+      container.inspect(function(error, stats) {
+        if ((error != null ? error.statusCode : undefined) === 404) {
+          return createAndStartContainer()
+        } else if (error != null) {
+          if (error.message.match(/EPIPE/)) {
+            if (!isRetry) {
+              metrics.inc('container-inspect-epipe-retry')
+              return inspectContainer(true)
+            }
+            metrics.inc('container-inspect-epipe-error')
+          }
+          logger.err(
+            { container_name: name, error },
+            'unable to inspect container to start'
+          )
+          return callback(error)
+        } else {
+          return startExistingContainer()
+        }
+      })
+    inspectContainer(false)
```

Comment:
A couple of thoughts:

- do we need a delay before retrying? (or more than one attempt to retry if  no delay is too short?)
- we have the async module (a very old version).  More recent versions of async include a `retry` method which allows a custom retry strategy: https://caolan.github.io/async/v3/docs.html#retry.  Maybe worth using that to standardise the retry approach?  Possibly we might find a few other things that need retries.

Reply:
Thanks for the comments!

From what I understand, the EPIPE errors are scoped to a bad connection and we open a new one for every request.
Looking [at the logs and high lighting the hostname](https://console.cloud.google.com/logs/query;query=unable%20to%20inspect%20container%20to%20start%0A;summaryFields=true,jsonPayload%252Fhostname:true:32:beginning?project=mgcp-1117973-ol-prod) these errors occur in clusters of long intervals (30s+). So slightly backing off may not help but increase the load on the box. Maybe we should mark the box as unavailable when seeing these errors?

Ideally the dockerode library would retry on EPIPE, but I could not find a config option and poking at the internals did not yield an easy hack either. https://github.com/apocas/docker-modem/blob/master/lib/http.js
Implementing retries in a transparent way is probably better in the long run?

Reply:
That's a good point; we thought about it but don't know for sure.
That's mostly why we added the two metrics: we'll have an easy way to monitor the effectiveness of the retries.
if `container-inspect-epipe-error` is close or equal to `container-inspect-epipe-retry` then we'll know the single, immediate retry doesn't address the issue.